### PR TITLE
Feature/tvist1 643 multiple complaint categories per case

### PR DESCRIPTION
**Pull request by jekuaitk**
https://jira.itkdev.dk/browse/TVIST1-643

* Allows multiple complaint categories per case
* Updates templates
* Updates logging logic
* Updates tests
* Adds `complaintCategories.names` macro to existing case macros

### CODE REVIEW

File: fixtures/case_entity.yaml [hunk 1]
```diff
@@ -412,7 +412,7 @@ App\Entity\ResidentComplaintBoardCase:
     hasVacated: false
     finishHearingDeadline: <(new \DateTime('2022-05-20'))>
     finishProcessingDeadline: <(new \DateTime('2022-07-11'))>
-    complaintCategory: '@security-deposit-complaint-aarhus'
+    complaintCategories: ['@security-deposit-complaint-aarhus']
```
[the remark this answers is not in the document]
Comment (jekuaitk) [reply fixtures/case_entity.yaml:415]: Added.

File: migrations/Version20221031135042.php [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+        $this->addSql('INSERT INTO case_entity_complaint_category (case_entity_id, complaint_category_id) SELECT id, complaint_category_id FROM case_entity');
+
+        $this->addSql('ALTER TABLE case_entity DROP FOREIGN KEY FK_A7C603C1D0DA653B');
+        $this->addSql('DROP INDEX IDX_A7C603C1D0DA653B ON case_entity');
+        $this->addSql('ALTER TABLE case_entity DROP complaint_category_id');
+    }
+
+    public function down(Schema $schema): void
+    {
+        // this down() migration is auto-generated, please modify it to your needs
```
[the remark this answers is not in the document]
Comment (jekuaitk) [reply migrations/Version20221031135042.php:39]: Added.

File: src/Service/MailTemplate/ComplexMacroHelper.php [hunk 3]
```diff
@@ -92,6 +94,20 @@ private function buildCaseMacros(CaseEntity $case): array
...
+        // Complaint categories comma separated names
+        $text = new Text();
+
+        $value = implode(', ', array_map(static function (ComplaintCategory $complaintCategory) {
+            return $complaintCategory->getName();
+        }, $case->getComplaintCategories()->toArray()));
+
+        $text->setText($value);
+
+        $values['complaintCategories.name'] = new ComplexMacro(
```
[the remark this answers is not in the document]
Comment (jekuaitk) [reply src/Service/MailTemplate/ComplexMacroHelper.php:106]: Good idea. Updated.

File: src/Form/AgendaCaseItemNewType.php [hunk 4]
```diff
@@ -40,13 +40,21 @@ public function buildForm(FormBuilderInterface $builder, array $options)
...
-                    $complaint = $caseEntity->getComplaintCategory()->getName();
+
+                    $complaints = '';
+                    foreach ($caseEntity->getComplaintCategories() as $complaintCategory) {
+                        if (empty($complaints)) {
+                            $complaints .= $complaintCategory->getName();
+                        } else {
+                            $complaints .= ', '.$complaintCategory->getName();
+                        }
+                    }
```
[the remark this answers is not in the document]
Comment (jekuaitk) [reply src/Form/AgendaCaseItemNewType.php:51]: Used `array_map` and `implode`.

File: src/Logging/EntityListener/AbstractEntityListener.php [hunk 5]
```diff
@@ -44,6 +45,40 @@ public function createLogEntry(string $action, CaseEntity $case, LifecycleEventA
...
+            /** @var $collectionUpdate PersistentCollection */
+            $removals = $collectionUpdate->getDeleteDiff();
+            $removalsArray = [];
+            $count = 1;
+            foreach ($removals as $removal) {
+                if ($removal instanceof ComplaintCategory) {
+                    $removalsArray[$count] = $removal->getName();
+                    ++$count;
+                }
+            }
```
[the remark this answers is not in the document]
Comment (jekuaitk) [reply src/Logging/EntityListener/AbstractEntityListener.php]: Updated.

File: src/Logging/EntityListener/AbstractEntityListener.php [hunk 6]
```diff
@@ -50,32 +50,27 @@ public function createLogEntry(string $action, CaseEntity $case, LifecycleEventA
...
-            foreach ($removals as $removal) {
-                if ($removal instanceof ComplaintCategory) {
-                    $removalsArray[$count] = $removal->getName();
-                    ++$count;
-                }
-            }
 
-            if ($removalsArray) {
-                $dataArray['Complaint category removals'] = $removalsArray;
+            $complaintCategoryRemovals = array_map(static function (ComplaintCategory $complaintCategory) {
```
Comment (rimi-itk) [inline src/Logging/EntityListener/AbstractEntityListener.php:54]: Umm, “arrow functions” …

**Review by rimi-itk — CHANGES_REQUESTED** [review]
One stylistic change needed. Consider using arrow functions.

File: migrations/Version20221031135042.php [hunk 7]
```diff
@@ -37,9 +37,13 @@ public function up(Schema $schema): void
     public function down(Schema $schema): void
     {
         // this down() migration is auto-generated, please modify it to your needs
-        $this->addSql('DROP TABLE case_entity_complaint_category');
         $this->addSql('ALTER TABLE case_entity ADD complaint_category_id BINARY(16) NOT NULL COMMENT \'(DC2Type:uuid)\'');
+
+        // Migrate just one of the complaint categories.
+        $this->addSql('UPDATE case_entity SET complaint_category_id = (SELECT complaint_category_id FROM case_entity_complaint_category WHERE case_entity_id = case_entity.id limit 1)');
```
[the remark this answers is not in the document]
Comment (jekuaitk) [reply migrations/Version20221031135042.php]: Fixed.

File: src/Form/AgendaCaseItemNewType.php [hunk 8]
```diff
@@ -41,14 +42,10 @@ public function buildForm(FormBuilderInterface $builder, array $options)
...
 
-                    $complaints = '';
-                    foreach ($caseEntity->getComplaintCategories() as $complaintCategory) {
-                        if (empty($complaints)) {
-                            $complaints .= $complaintCategory->getName();
-                        } else {
-                            $complaints .= ', '.$complaintCategory->getName();
-                        }
-                    }
+                    $complaints = implode(', ', array_map(static function (ComplaintCategory $complaintCategory) {
```
[the remark this answers is not in the document]
Comment (jekuaitk) [reply src/Form/AgendaCaseItemNewType.php]: Increased readability via arrow functions.

File: src/Logging/EntityListener/AbstractEntityListener.php (hunk 6 above)
Comment (jekuaitk) [reply src/Logging/EntityListener/AbstractEntityListener.php]: Increased readability via arrow functions.

**Review by rimi-itk — APPROVED** [review]

**Pull request merged**